### PR TITLE
[feature] support configuring database caches

### PR DESCRIPTION
**Pull request by NyaaaWhatsUpDoc**
# Description

- adds `CacheConfiguration{}` options to main configuration struct, allowing setting via config file
- update config helpers generator to support nested structs
- add example cache config variables to config tests

## Checklist

- [x] I/we have read the [GoToSocial contribution guidelines](https://github.com/superseriousbusiness/gotosocial/blob/main/CONTRIBUTING.md).
- [x] I/we have discussed the proposed changes already, either in an issue on the repository, or in the Matrix chat.
- [x] I/we have performed a self-review of added code.
- [x] I/we have written code that is legible and maintainable by others.
- [x] I/we have commented the added code, particularly in hard-to-understand areas.
- [x] I/we have made any necessary changes to documentation.
- [x] I/we have added tests that cover new code.
- [x] I/we have run tests and they pass locally with the changes.
- [x] I/we have run `go fmt ./...` and `golangci-lint run`.

### CODE REVIEW

**Comment by NyaaaWhatsUpDoc** [comment]
Only thing still remaining is to update example configuration file, and possibly update other areas of documentation?

**Comment by NyaaaWhatsUpDoc** [comment]
In time I want to add intelligent default for cache sizes.

**Comment by NyaaaWhatsUpDoc** [comment]
This lays the groundwork for any changes (if we decide to make them) in #1244

**Comment by tsmethurst** [comment]
Nice one kim! this looks really good :)

We gotta remember when we cut a release including these changes to include a migration note about the config format changing to allow nesting.

Also, do you foresee any issues with the nesting? I mean, could it break if one provides a config that doesn't nest?

**Comment by NyaaaWhatsUpDoc** [comment]
> Nice one kim! this looks really good :)
> 
> We gotta remember when we cut a release including these changes to include a migration note about the config format changing to allow nesting.
> 
> Also, do you foresee any issues with the nesting? I mean, could it break if one provides a config that doesn't nest?

for this specific PR there shouldn't be any breaking changes. the only nested values are new ones so shouldn't be an issue. for later PRs though -- I added a note about this in the issue #1244 :)

File: internal/config/state.go [hunk 1]
```diff
@@ -41,8 +41,8 @@ func NewState() *ConfigState {
 	viper := viper.New()
 
 	// Flag 'some-flag-name' becomes env var 'GTS_SOME_FLAG_NAME'
-	viper.SetEnvPrefix("gts")
 	viper.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
+	viper.SetEnvPrefix("gts")
```
Comment (tsmethurst) [inline internal/config/state.go:45]: Is the order important here? If so, worth commenting it imo

**Review by tsmethurst — APPROVED** [review]
Great :) Just the one comment

File: internal/config/state.go (hunk 1 above)
Comment (NyaaaWhatsUpDoc) [reply internal/config/state.go:45]: oh no importance, was literally just a "this looks better this way round :')"

I have very neurotypical brain

**Pull request merged**